### DISCUSSION TITLE
Swapped to containerd from docker and disk usage has doubled - how to fix?

### DISCUSSION THREAD

**Discussion by daniel-barron-mk** [k3s-io/k3s#9971] [Q&A]
k3s 1.29.3.

All comparisons are on two freshly installed identical CentOS VMs. The only difference is the docker/containerd setting.

I was using `--docker` when starting k3s, coupled with `docker load`. Testing performance I found that not using `--docker` (and thus containerd) with `/usr/local/bin/k3s ctr images import` gave me a needed small performance increase. Also containerd seems to be the standard nowadays.

What I've found is that `/var/lib/rancher/k3s/agent/containerd` is taking twice the space that `/var/lib//var/lib/docker/overlay2` used to take.

Using docker:
```
00:00:44  Filesystem                      Size  Used Avail Use% Mounted on
00:00:44  /dev/mapper/vg_main-lv_var     20G  8.7G   10G  47% /var
```
Using containerd:
```
00:00:49  /dev/mapper/vg_main-lv_var     20G   17G  2.3G  88% /var
```

Using `du` on /var/lib/ is about 17G on either. But I think du is counting things twice. Using `du -x` shows docker at 8G and containerd at 14G.

This doubling of space is a problem where I am using k3s where I can't increase the disk space.

Any ideas please?

**Comment by brandond** [comment]
You might try allowing the image store to discard the unpacked layers?

You'll need to provide your own containerd.tomp.tmpl, as described in the docs: 
* https://docs.k3s.io/advanced#configuring-containerd
* https://github.com/k3s-io/k3s/blob/d973fadbed133cf9c5db64c87ae724b1a80c766f/pkg/agent/templates/templates_linux.go#L10-L119

Add a line in the `[plugins."io.containerd.grpc.v1.cri".containerd]` table, setting `discard_unpacked_layers = true`

Ref: https://github.com/containerd/containerd/blob/8317959018015f6a1756ec8cd08be1093fd630a2/docs/cri/config.md?plain=1#L279-L281

**Comment by daniel-barron-mk** [comment]
Thanks for the idea. Finally got round to it today while trying other idea before this. Unfortunately it makes no difference at all.

Example numbers - size of /var/lib/rancher/:

Using docker with Docker image format files: **9 GB**
Using containerd with Docker image format files: **19 GB**
Using containerd with Docker image format files with discard_unpacked_layers = true: **19 GB**
Using containerd with OCI image format files: **12 GB** (better but not good enough)

Any other ideas to get the disk space down to the same level as docker please?

**Reply by brandond** [reply]
This is all just containerd behavior, not k3s. You might take your inquiry to the containerd project.

**Reply by daniel-barron-mk** [reply]
Yes you are right. I wondered if it was specific to k3s containerd.

Anyway I have an update.

By putting in a go template to change the config for containerd to `discard_unpacked_layers = true` AND also ensuring the images were in OCI image format the size is back down to 9 GB.

However an even simpler answer was to add:
unzstd $image -c | /usr/local/bin/k3s ctr images import <b>--all-platforms --discard-unpacked-layers</b> -

**Reply by brandond** [reply]
Settings from the `plugins."io.containerd.grpc.v1.cri".containerd` section are specific to the CRI service only take effect when using the CRI interface (kubelet or crictl) to interact with containerd - if you're using `ctr` to import images then yes, you'd need to manually set flags to discard unpacked layers.
